modules/sdk-core/src/bitgo/lightning/validateLnurlInvoice.ts

```diff
@@ -0,0 +1,63 @@
+import * as bolt11 from 'bolt11';
+import * as crypto from 'crypto';
+
+export type ParsedLightningInvoice = {
+  amount: number;
+  /** hex-encoded */
+  paymentHash: string;
+  /** hex-encoded */
+  pubkey: string;
+};
+
+function getNetworkForInvoice(invoice: string) {
+  if (invoice.startsWith('lntbs')) {
+    // signet invoices are not supported by our bolt11 dependency
+    // because the `tbs` prefix used for the invoice does not match the `tb`
+    // prefix used by on-chain signet addresses
+    // see: https://github.com/bitcoinjs/bolt11/pull/58#issuecomment-1106495709
+    // we can still decode the invoice however using a custom network
+    return {
+      bech32: 'tbs',
+      pubKeyHash: 0x6f,
+      scriptHash: 0xc4,
+      validWitnessVersions: [0, 1],
+    };
+  }
+
+  return undefined;
+}
+
+export function validateLnurlInvoice(coin: string, invoiceStr: string, milliSatAmount: string, metadata: string): void {
+  if (coin === 'tbtc') {
+    // tb = testnet invoice, crt = regtest invoice
+    if (!invoiceStr.startsWith('lntb') && !invoiceStr.startsWith('lncrt') && !invoiceStr.startsWith('lntbs')) {
+      throw new Error('invoice is not for tbtc');
+    }
+  } else {
+    if (!invoiceStr.startsWith('lnbc')) {
+      throw new Error('invoice is not for mainnet btc');
+    }
+  }
+
+  let decodedInvoice;
+  try {
+    decodedInvoice = bolt11.decode(invoiceStr, getNetworkForInvoice(invoiceStr));
+  } catch (e) {
+    throw new Error(`invoice is invalid`);
+  }
+
+  const { millisatoshis, tags } = decodedInvoice;
+  if (millisatoshis === null || millisatoshis === undefined) {
+    throw new Error('invoice does not have an amount');
+  }
+
+  if (millisatoshis !== milliSatAmount) {
+    throw new Error('amount of invoice does not match with given amount');
+  }
+
+  const hash = crypto.createHash('sha256').update(new TextEncoder().encode(metadata)).digest('hex');
+  const found = tags.find((tag) => tag.data === hash);
```

Comment:
Rather than finding by `data`, I think it's better to find by `tagName === 'purpose_commit_hash'` and then check its value. So we don't accidentally pass it if the right value is in the wrong tag.

```suggestion
  const descriptionHash = tags.find((tag) => tag.tagName === 'purpose_commit_hash');
  if (descriptionHash !== hash)
```

Reply:
for `domain`, type is `t.union([t.string, t.undefined)` because it `LnurlPayResponse` is first used to decode the response from ln service, which does not contain `domain` field, and later we manually add `domain` to the final return value. I have done this way so that I do not have to create another type. Or actually we can omit `domain` in `LnUrlPayResponse` and make return type be like `LnurlPayResponse & { domain: string }`